src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Folding/FoldingRangeEndpoint.cs

```diff
@@ -92,6 +93,57 @@ public TextDocumentIdentifier GetTextDocumentIdentifier(FoldingRangeParams reque
         return foldingRanges;
     }
 
+    private class RazorFileFoldingRangeSyntaxWalker : SyntaxWalker
+    {
+        private readonly RazorSourceDocument _source;
+        internal List<FoldingRange> Ranges { get; }
+
+        public RazorFileFoldingRangeSyntaxWalker(RazorSourceDocument source)
+        {
+            _source = source;
+            Ranges = new List<FoldingRange>();
+        }
+
+        public override void VisitRazorDirective(RazorDirectiveSyntax node)
+        {
+            // first rule: we can fold consecutive usings
+            // if we have an existing range that ends the line *before* this current using directive or begins the line *after*,
+            // we extend the range one line
+            if (node.Body is RazorDirectiveBodySyntax syntax && string.Equals(syntax.Keyword.GetFirstToken().Content, "using"))
+            {
+                var linePosition = node.GetLinePositionSpan(_source);
+
+                var isPartOfExistingRange = false;
+                foreach (var range in Ranges)
+                {
+                    if (range.EndLine + 1 == linePosition.Start.Line)
+                    {
+                        range.EndLine = linePosition.End.Line;
+                        range.EndCharacter = linePosition.End.Character;
+                        isPartOfExistingRange = true;
+                    }
+                    else if (range.StartLine - 1 == linePosition.End.Line)
+                    {
+                        range.StartLine = linePosition.Start.Line;
+                        range.StartCharacter = linePosition.Start.Character;
+                        isPartOfExistingRange = true;
+                    }
+                }
+
+                if (!isPartOfExistingRange)
+                {
+                    Ranges.Add(new FoldingRange
```

Comment:
According to [the spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_foldingRange) there is a folding range kind for "imports". No idea if any clients support it, but we should probably support it, if the `FoldingRange` type exposes it